src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php

```diff
@@ -56,9 +56,9 @@ public function __construct(TokenStorageInterface $tokenStorage, EventDispatcher
     /**
      * Authenticates the given token in the system.
      */
-    public function authenticateWithToken(TokenInterface $token, Request $request, string $providerKey = null)
+    public function authenticateWithToken(TokenInterface $token, Request $request, string $providerKey = null, TokenInterface $previousToken = null)
     {
-        $this->migrateSession($request, $token, $providerKey);
+        $this->migrateSession($request, $token, $providerKey, 3 < \func_num_args() ? $previousToken : $this->tokenStorage->getToken());
```

Comment:
This class is `@final` so adding the arg is nt a BC break.

For extra safety, I ensured the previous token should always be read *before* calling `authenticate()`. In case the caller hasn't been updated to the new signature, this check fallback to reading the previous token here.

Reply:
Is there something against not adding this argument and doing `$this->tokenStorage->getToken()` in `migrateSession()`? That seems a bit safer BC wise, given this is an LTS release.

Reply:
`$this->tokenStorage->getToken()` should be called before calling `authenticate()`, that's why.
I still want a fallback path because getting a proper previous token requires upgrading both security-bundle (to wire the new argument)